Geometry_Engine/Query/GeometryHash.cs

```diff
@@ -0,0 +1,409 @@
+        {
+            return GeometryHash(obj as dynamic, translationFactor);
+        }
+
+        /***************************************************/
+
+        [Description("The GeometryHash for a Line is calculated as the GeometryHash of the External and Internal boundary curves, then concatenated.")]
+        private static double[] GeometryHash(this PlanarSurface obj, double translationFactor)
+        {
+            translationFactor += (int)TypeTranslationFactor.PlanarSurface;
+
+            return obj.ExternalBoundary.GeometryHash(translationFactor)
+                .Concat(obj.InternalBoundaries.SelectMany(ib => ib.GeometryHash(translationFactor))).ToArray();
+        }
+
+        /***************************************************/
+
+        [Description("The GeometryHash for an Extrusion is calculated by translating the extrusion curve with the extrusion direction vector." +
+            "A first GeometryHash is calculated for this translated curve. " +
+            "Then, the GeometryHash of the (non-translated) extrusion curve is concatenated to the first hash to make it more reliable.")]
+        private static double[] GeometryHash(this Extrusion obj, double translationFactor)
+        {
+            translationFactor += (int)TypeTranslationFactor.Extrusion;
+
+            return obj.Curve.ITranslate(obj.Direction).GeometryHash(translationFactor)
+                .Concat(obj.Curve.GeometryHash(translationFactor)).ToArray();
+        }
+
+        /***************************************************/
+
+        [Description("The GeometryHash for a Loft is calculated as the GeometryHash of its curves.")]
+        private static double[] GeometryHash(this Loft obj, double translationFactor)
+        {
+            translationFactor += (int)TypeTranslationFactor.Loft;
+
+            return obj.Curves.GeometryHash(translationFactor);
+        }
+
+        /***************************************************/
+
+        [Description("Moving control points by a translation factor composed by the weights " +
+          "and a subarray of the knot vector. " +
+          "The subarray is made by picking as many elements from the knot vector as the curve degree value.")]
+        private static double[] GeometryHash(this NurbsSurface obj, double translationFactor)
+        {
+            translationFactor += (int)TypeTranslationFactor.NurbsSurface;
+
+            List<double> concatenated = new List<double>();
+            for (int i = 0; i < obj.ControlPoints.Count(); i++)
+            {
+                double UKnotsSum = obj.UKnots.ToList().GetRange(i, obj.UDegree).Sum();
+                double VKnotsSum = obj.VKnots.ToList().GetRange(i, obj.VDegree).Sum();
+
+                double[] doubles = obj.ControlPoints[i].GeometryHash(UKnotsSum + VKnotsSum + obj.Weights[i] + translationFactor);
+                concatenated.AddRange(doubles);
+            }
+
```

Comment:
Think we need to handle the SurfaceTrims here as well?

Reply:
Good catch!